# Laboratorio de regresión - 3

## Significancia de factores

|                |   |
:----------------|---|
| **Nombre**     | Adrian Marcelo Ballesteros Herrera  |
| **Fecha**      |  1/09/2025 |
| **Expediente** | 750743 |

Descarga el archivo de publicidad y carga los datos (Advertising.csv).

In [64]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression

### Anotaciones marcelo

errores comunes al hacer una regresion lineal *Overfitting* y *Overflow*. el overfitting es cuando haces una regresion de demasiados grados para que de una $R^2$ muy baja pero no expresa correctamente los datos y el overflow es cuando se te rompe el codigo por que los numeros son muy altos por la matriz de grados muy altos.

In [65]:
datos=pd.read_csv(r'D:\Advertising.csv')

In [66]:
datos.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

**¿Hay alguna relación entre el presupuesto para publicidad y las ventas?**

Nuestra primera meta debe ser determinar si hay evidencia en los datos de que haya una asociación entre estas variables.

- ¿Por qué? ¿Qué resultaría si nos diéramos cuenta de la falta de relación entre el presupuesto de publicidad y las ventas?

podemos ver si vale la pena usar una regresion o no, si vemos el coeficiente o los coeficientes y son muy altos, no podemos asumir que son distintos de 0 por lo que tenemos que hacer una prueba de hipotesis para ya despues de eso hacer un analisis de como incrementar las ventas y en que tipo de publicidad influye mas.

**¿Qué tan fuerte es esta relación?**
Asumiendo que existe esta relación, ¿nos sirve conocer el impacto que tiene invertir en publicidad en las ventas?

Claro, uno tiene que buscar siempre como mejorar las ventas y asi nos gastamos muchisimo menos dinero y no perdemos dinero. Por el otro lado si la relacion es muy fuerte y tenemos buenos coeficientes de $B1$, tenemos que invertirle mas dinero para conseguir mas clientes, ya luego retenerlos es otra cosa

**¿Cuáles medios están asociados con las ventas? ¿Qué tan grande es la asociación entre un medio específico y las ventas?**

Hay 3 medios distintos en los datos. ¿Sirve invertir en los 3? ¿Conviene más invertir sólo en uno?

**¿Qué tan seguros estamos de que podríamos predecir ventas futuras?**

**¿La relación es lineal?**

**¿Hay sinergia entre estos medios?**

Puede ser que gastar \\$50,000 en publicidad y otros \\$50,000 en radio es mejor opción que gastar \\$100,000 en televisión. A esto le llamamos efecto de interacción.

Podemos usar regresión lineal para responder todas estas preguntas.

Realiza una regresión lineal:

$$ \text{ventas} \approx \beta_0 + (\beta_1)(\text{TV})$$

In [67]:
X=datos["TV"].values.reshape(-1,1)
x1=datos['newspaper'].values.reshape(-1,1)
x2=datos['radio'].values.reshape(-1,1)
y=datos['sales']

In [68]:
xfull=np.hstack([X,x1,x2])
lr=LinearRegression ()
lr.fit(X,y)

LinearRegression()

In [69]:
print(lr.coef_)
print(lr.intercept_)
print(lr.score(X,y))

[0.04753664]
7.032593549127693
0.611875050850071


### Verificando la precisión de nuestros coeficientes estimados

Recuerda que en el mundo real hay ruidos y errores de medición. Siempre se asume que la verdadera relación entre $X$ y $Y$ es $$Y = \beta_0 + \beta_1 X + \epsilon$$

Se asume que el término de error es independiente de $X$ (el error siempre es el mismo sin importar el valor de $X$). Este modelo describe a la *línea de regresión de la población*, que es la mejor aproximación de la verdadera relación entre $X$ y $Y$. Cuando usamos mínimos cuadrados encontramos la *línea de mínimos cuadrados*.

¿Cuál es la diferencia entre población y muestra?

que la poblacion es todo el mundo, la muestra es una pequeña parte de la prueba y dependiendo de si calculamos bien nuestra muestra, podremos sacar un valor parecido a la poblacion aun que hay un error por que no tenemos absolutamente todos los datos.

¿Cuál crees que sea la diferencia entre hacer una regresión con todos los datos de la población y una muestra de ella?

Que al hacer una regresion con todos los datos, no hay error y es una regresion total, en la funcion de y=f(x) + $E$ se elimina y es una regresion sin fallas, ahora al hacer una muestra, en cada dato se hace una distribucion normal de los datos que dependiendo de la desviacion estandar de la regresion y que tan alejados estan de la recta

La línea de regresión de la población no se puede observar. El concepto de comparar estas líneas es una extensión natural del acercamiento estadístico estándar de usar información de una muestra para estimar características de una población grande.

Imagina que quieres encontrar la altura promedio de un mexicano $\mu$. Medir a todos y cada uno de los mexicanos en situaciones similares, con la misma regla, mismo operador, y otras incontables formas de minimizar la variación de la medida es una tarea imposible. Lo que podemos asumir es que $\hat{\mu} = \bar{y}$. La media poblacional y la media muestral son diferentes, pero la media muestral es usualmente un buen estimado.

De la misma manera, como no contamos con el 100% de la información para hacer una regresión, los coeficientes $\beta_0$ y $\beta_1$ son desconocidos. Podemos estimarlos usando mínimos cuadrados, encontrando $\hat{\beta_0}$ y $\hat{\beta_1}$. Puede que las muestras que tengamos en ese momento estén un poco por encima de la media, pero otras muestras en otro momento puede que estén debajo de la media. En general, esperamos que el promedio de las aproximaciones $\hat{\mu}$ aproxime a $\mu$.

Esto lleva a la pregunta: ¿qué tan cercanos son nuestros coeficientes estimados a los verdaderos coeficientes? Utilizamos el concepto de error estándar para evaluar esto.

$$ \text{Var}(\hat{\mu})=\text{SE}(\hat{\mu})^2 = \frac{\sigma^2}{n} $$

Donde $\sigma$ es la desviación estándar de cada una de las observaciones $y_i$ de $Y$. El error estándar nos dice la cantidad promedio que el estimado difiere del valor verdadero. Podemos ver en la fórmula que entre más observaciones tengamos el error se hace más pequeño. Las fórmulas para errores estándar de $\hat{\beta_0}$ y $\hat{\beta_1}$ son:

$$ \text{SE}(\hat{\beta_0})^2 = \sigma^2 [\frac{1}{n} + \frac{\bar{x}^2}{\sum_{i=1}^n (x_i - \bar{x})^2}]$$

$$ \text{SE}(\hat{\beta_1})^2 = \frac{\sigma^2}{\sum_{i=1}^n (x_i - \bar{x})^2}$$

$$ \sigma^2 = \text{Var}(\epsilon) = \text{RSE}^2 = \frac{\text{RSS}}{n-p}$$

Para que estas fórmulas sean validas asumimos que los errores $\epsilon_i$ tienen varianza común $\sigma^2$ y que no están correlacionados.

Calcula los errores estándar de los coeficientes

In [70]:
xM=np.mean(X)
yM=np.mean(y)
n=len(y)
TSS=np.sum((y-yM)**2) 
RSS=abs((lr.score(X,y)-1)*TSS)
yV=abs(RSS)/(n-2)
SE02=yV*(1/n + (xM**2)/np.sum((X-xM)**2))
SE12=yV/np.sum((X-xM)**2)

In [71]:
print(TSS)
print(RSS)
print(yV)
SE0= np.sqrt(SE02)
SE1=np.sqrt(SE12)
print(SE0)
print(SE1)

5417.14875
2102.5305831313517
10.618841328946221
0.45784294027347855
0.0026906071877968707


Estos errores se pueden usar para calcular intervalos de confianza. Un intervalo de confianza del $95\%$ se define como un rango de valores en el cuál se encuentra el desconocido valor verdadero con un $95\%$ de probabilidad.

Otra forma de verlo es que si tomamos muestras repetidas y construimos un intervalo de confianza para cada una, el $95\%$ de los intervalos creados van a contener el valor verdadero. Para la regresión el intervalo de confianza del $95\%$ toma la forma:

$$ \hat{\beta_j} \pm 2\text{SE}(\hat{\beta_j})$$

Calcula los intervalos de confianza para los coeficientes estimados:

In [72]:
Boin_upper=lr.intercept_+2*SE0
Boin_lower=lr.intercept_-2*SE0

In [73]:
B1in_upper=lr.coef_+(2*SE1)
B1in_lower=lr.coef_-(2*SE1)

In [74]:
print(Boin_upper,Boin_lower,
B1in_upper,B1in_lower)

7.94827942967465 6.116907668580736 [0.05291785] [0.04215543]


Los errores estándar también se usan para realizar pruebas de hipótesis. La prueba de hipótesis más común es probar la hipótesis nula de:

$$ H_0: \text{No hay relación entre } X \text{ y } Y \ \ \ \ (\beta_1=0)$$

contra la hipótesis alternativa:
$$ H_A: \text{Hay alguna relación entre } X \text{ y } Y \ \ \ (\beta_1 \neq 0)$$

Explica con tus palabras el significado de la hipótesis nula y la hipótesis alternativa.

Cuando hacemos una prueba de hipotesis, uno siempre busca rechazar la hipotesis nula que en este caso es mostrar que si es estadisticamente significativo el valor de B1 o de otra forma, distinto de 0

Para probal la hipótesis nula debemos determinar si nuestro estimado $\hat{\beta_1}$ de $\beta_1$ está lo suficientemente alejado de cero para que podamos decir con confianza que este valor no es cero. 

¿Qué tan lejos? Depende de qué tanta confianza tengamos en el estimado encontrado. Si nuestro error estándar es pequeño y nuestro estimado está alejado de cero podríamos decir que hay muy poca probabilidad de que el valor verdadero sea 0. En cambio, si nuestro error estándar es grande y nuestro estimado está muy cerca de cero, entonces podrías ser que el valor verdadero sea cero y que no haya relación entre las variables.

Se calcula un *estadístico t* dado por
$$ t = \frac{\hat{\beta_j} - \mu}{\text{SE}(\hat{\beta_j})} $$

donde $\mu$ es el valor contra el que queremos probar.

Calcula el estadístico t para tus coeficientes estimados, usando como referencia la prueba de hipótesis.

In [75]:
t=lr.coef_/SE1
print(t)

[17.6676256]


La distribución t tiene forma de campana y se parece bastante a la distribución normal cuando $n > 30$. Ya sólo es cuestión de calcular la probabilidad de observar cualquier número tal que su valor absoluto sea igual o mayor que el valor absoluto del estadístico t calculado. En otras palabras:
$$ P(|x| \geq |t|) $$

A esta probabilidad la llamamos *p-value*. Un *p-value* pequeño indica que es poco probable que exista por puro azar una relación significativa entre predictor y respuesta, en caso de que no haya una asociación real entre predictor y respuesta. En otras palabras, el *p-value* te dice la probabilidad de que parezca que hay relación cuando no la hay.

Si el *p-value* es pequeño, inferimos que sí hay una asociación entre el predictor y la respuesta, y **rechazamos la hipótesis nula**.
  

¿Qué tan pequeño? Depende de la aplicación. Un valor muy común es del $5\%$.

Utiliza el siguiente código para calcular el *p-value* para tus coeficientes

`from scipy import stats`

`p_bj = 2*(1 - stats.t.cdf(np.abs(t_bj), n-p))`

In [76]:
from scipy import stats
p_bj=2*(1-stats.t.cdf(np.abs(t),n-1))

In [77]:
print(p_bj,
yM,
yV)

[0.] 14.0225 10.618841328946221


¿Se rechaza la hipótesis nula? ¿Qué significa?

el valor estimado esta alejado de 0 y nuestra prueba de t nos da casi 0 por lo que no rechazamos la hipotesis nula, esto significa que podemos asumir con un 90% de confianza que B0 es igual a 0  

Realiza otras dos regresiones. Ya tienes hecha la regresión de ventas dado el gasto en publicidad de TV. Realiza la regresión para gastos en radio y gastos en periódico. Organiza las respuestas para que debajo de esta celda se tenga:
- Título de regresión
- Coeficientes estimados
- Errores estándar de los coeficientes
- Intervalos de confianza
- Estadísticos t
- p-values
- Observaciones

## Esta es la regresion de las ventas con la publicidad de la radio

In [78]:
x=np.hstack([x1])
lr.fit(x,y)
print(lr.coef_)
print(lr.intercept_)
print(lr.score(x,y))

[0.0546931]
12.35140706927816
0.05212044544430516


In [79]:
xM=np.mean(X)
yM=np.mean(y)
n=len(y)
TSS=np.sum((y-yM)**2) 
RSS=abs((lr.score(X,y)-1)*TSS)
yV=abs(RSS)/(n-2)
SE02=yV*(1/n + (xM**2)/np.sum((X-xM)**2))
SE12=yV/np.sum((X-xM)**2)

In [80]:
print(TSS)
print(RSS)
print(yV)
SE0= np.sqrt(SE02)
SE1=np.sqrt(SE12)
print(SE0)
print(SE1)

5417.14875
10295.879929858309
51.999393585142975
1.0131578079771122
0.005954027988042014


In [81]:
Boin_upper=lr.intercept_+2*SE0
Boin_lower=lr.intercept_-2*SE0
B1in_upper=lr.coef_+(2*SE1)
B1in_lower=lr.coef_-(2*SE1)
print(Boin_upper,Boin_lower,
B1in_upper,B1in_lower)

14.377722685232385 10.325091453323935 [0.06660115] [0.04278504]


In [82]:
t=lr.coef_/SE1
print(t)
from scipy import stats
p_bj=2*(1-stats.t.cdf(np.abs(t),n-1))
print(p_bj,
yM,
yV)

[9.18589879]
[0.] 14.0225 51.999393585142975


## y por ultimo la regresion de las ventas con el periodico 

In [83]:
x=np.hstack([x2])
lr.fit(x,y)
print(lr.coef_)
print(lr.intercept_)
print(lr.score(x,y))

[0.20249578]
9.311638095158283
0.33203245544529525


In [84]:
xM=np.mean(X)
yM=np.mean(y)
n=len(y)
TSS=np.sum((y-yM)**2) 
RSS=abs((lr.score(X,y)-1)*TSS)
yV=abs(RSS)/(n-2)
SE02=yV*(1/n + (xM**2)/np.sum((X-xM)**2))
SE12=yV/np.sum((X-xM)**2)

In [85]:
print(TSS)
print(RSS)
print(yV)
SE0= np.sqrt(SE02)
SE1=np.sqrt(SE12)
print(SE0)
print(SE1)

5417.14875
162971.36044026565
823.0876789912406
4.030890555292411
0.023688348442839426


In [86]:
Boin_upper=lr.intercept_+2*SE0
Boin_lower=lr.intercept_-2*SE0
B1in_upper=lr.coef_+(2*SE1)
B1in_lower=lr.coef_-(2*SE1)
print(Boin_upper,Boin_lower,
B1in_upper,B1in_lower)

17.373419205743104 1.2498569845734622 [0.24987248] [0.15511909]


In [87]:
t=lr.coef_/SE1
print(t)
from scipy import stats
p_bj=2*(1-stats.t.cdf(np.abs(t),n-1))
print(p_bj,
yM,
yV)

[8.54832847]
[3.33066907e-15] 14.0225 823.0876789912406


## Regresión lineal múltiple

En lugar de hacer una regresión para cada factor independiente, quizás se puede extender el modelo para que tenga varios factores dentro:

$$ Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_p X_p + \epsilon $$

Para nuestro ejemplo de publicidad:

$$ \text{sales} = \beta_0 + \beta_1 (\text{TV}) + \beta_2 (\text{radio}) + \beta_3 (\text{newspaper}) + \epsilon $$

Utiliza la librería `statsmodels` para realizar la regresión. Por defecto la librería statsmodels no toma en cuenta el intercepto ($\beta_0$), por lo que se tendrá que agregar una columna de unos de tamaño *n* a la matriz X.

`import statsmodels.api as sm`

`ols = sm.OLS(Y, X)`

`results = ols.fit()`

`results.summary()`

In [ ]:
ones=np.ones([n,1])
X=datos["TV"].values.reshape(-1,1)
x1=datos['newspaper'].values.reshape(-1,1)
x2=datos['newspaper'].values.reshape(-1,1)
xfull=np.hstack([ones,X,x1,x2])
y=datos['sales']
lr.fit(xfull,y)
print(lr.coef_)
print(lr.intercept_)
print(lr.score(xfull,y))

[ 0.          0.04576465 -0.00103749  0.18853002]
2.9388893694594103
0.8972106381789522


In [89]:
import statsmodels.api as sm
ols= sm.OLS(y,xfull)
results=ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Thu, 28 Aug 2025   Prob (F-statistic):           1.58e-96
Time:                        17:43:19   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
x1             0.0458      0.001     32.809      0.000       0.043       0.049
x2            -0.0010      0.006     -0.177      0.860      -0.013       0.011
x3             0.1885      0.009     21.893      0.000       0.172       0.206
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [90]:
ones=np.ones([n,1])
X=datos["TV"].values.reshape(-1,1)
x1=datos['radio'].values.reshape(-1,1)
xfull=np.hstack([ones,X,x1,x*x1])
y=datos['sales']
lr.fit(xfull,y)
print(lr.coef_)
print(lr.intercept_)
print(lr.score(xfull,y))
import statsmodels.api as sm
ols= sm.OLS(y,xfull)
results=ols.fit()
results.summary()

[0.         0.04576711 0.15600099 0.00066767]
3.155796590644222
0.8977739777342808


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.898
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     573.8
Date:                Thu, 28 Aug 2025   Prob (F-statistic):           9.20e-97
Time:                        17:43:19   Log-Likelihood:                -385.63
No. Observations:                 200   AIC:                             779.3
Df Residuals:                     196   BIC:                             792.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.1558      0.369      8.550      0.000       2.428       3.884
x1             0.0458      0.001     32.926      0.000       0.043       0.049
x2             0.1560      0.031      4.969      0.000       0.094       0.218
x3             0.0007      0.001      1.054      0.293      -0.001       0.002
==============================================================================
Omnibus:                       66.663   Durbin-Watson:                   2.075
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              173.446
Skew:                          -1.456   Prob(JB):                     2.17e-38
Kurtosis:                       6.513   Cond. No.                     3.31e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.31e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [91]:
x2=datos['newspaper'].values.reshape(-1,1)
xfull=np.hstack([ones,X,x1,x2, x1*x2,X*x1,x*x1])
lr.fit(xfull,y)
print(lr.coef_)
print(lr.intercept_)
print(lr.score(xfull,y))
import statsmodels.api as sm
ols= sm.OLS(y,xfull)
results=ols.fit()
results.summary()

[ 0.00000000e+00  1.90802845e-02  2.27935933e-02  2.92087996e-03
 -5.40569150e-05  1.08650187e-03  1.45552558e-04]
6.739028384867179
0.9678503333428412


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.968
Model:                            OLS   Adj. R-squared:                  0.967
Method:                 Least Squares   F-statistic:                     968.4
Date:                Thu, 28 Aug 2025   Prob (F-statistic):          3.94e-141
Time:                        17:46:24   Log-Likelihood:                -269.95
No. Observations:                 200   AIC:                             553.9
Df Residuals:                     193   BIC:                             577.0
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.7390      0.311     21.652      0.000       6.125       7.353
x1             0.0191      0.002     12.520      0.000       0.016       0.022
x2             0.0228      0.019      1.196      0.233      -0.015       0.060
x3             0.0029      0.008      0.372      0.710      -0.013       0.018
x4         -5.406e-05      0.000     -0.221      0.825      -0.001       0.000
x5             0.0011    5.3e-05     20.492      0.000       0.001       0.001
x6             0.0001      0.000      0.391      0.696      -0.001       0.001
==============================================================================
Omnibus:                      126.945   Durbin-Watson:                   2.209
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1150.750
Skew:                          -2.302   Prob(JB):                    1.31e-250
Kurtosis:                      13.812   Cond. No.                     2.32e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.32e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### ¿Qué diferencias puedes observar entre los *p-values* de una regresión múltiple y los encontrados en las regresiones simples? ¿Por qué crees que existen estas diferencias?

Al tener mas informacion, nuestro analisis es mucho mas amplio y nos da informacion que haciendo regresiones con cada x no nos daba, ahora podemos ver que el periodico es insignificante por que nuestro $p-values$ sale de 0.86 pero los demas si son distintos de 0 y si son estadisticamente significativos para nuestra regresion.

## Referencia

James, G., Witten, D., Hastie, T., Tibshirani, R.,, Taylor, J. (2023). An Introduction to Statistical Learning with Applications in Python. Cham: Springer. ISBN: 978-3-031-38746-3

## Addendum

Para calcular los *p-values* de los parámetros sin `statsmodels`:
1. Calcular RSS
2. Calcular RSE
3. `var_beta = np.linalg.inv(X.T @ X) * rse**2` (X con columna de intercepto)
4. `std_beta = np.sqrt(var_beta.diagonal())` El orden de los valores corresponde al orden de los factores en las columnas de la matriz $X$.
5. Calcular *estadístico t*
6. Calcular *p-value*